genie-web/src/main/java/com/netflix/genie/web/services/impl/ArchivedJobServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Set;
+
+/**
+ * Default implementation of {@link ArchivedJobService}.
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@Slf4j
+public class ArchivedJobServiceImpl implements ArchivedJobService {
+    private static final String SLASH = "/";
+    private static final String GET_ARCHIVED_JOB_METADATA_METRIC_NAME
+        = "genie.web.services.archivedJobService.getArchivedJobMetadata.timer";
+
+    private final JobPersistenceService jobPersistenceService;
+    private final ResourceLoader resourceLoader;
+    private final MeterRegistry meterRegistry;
+
+    /**
+     * Constructor.
+     *
+     * @param jobPersistenceService The {@link JobPersistenceService} implementation to use
+     * @param resourceLoader        The {@link ResourceLoader} used to get resources
+     * @param meterRegistry         The {@link MeterRegistry} used to collect metrics
+     */
+    public ArchivedJobServiceImpl(
+        final JobPersistenceService jobPersistenceService,
+        final ResourceLoader resourceLoader,
+        final MeterRegistry meterRegistry
+    ) {
+        this.jobPersistenceService = jobPersistenceService;
+        this.resourceLoader = resourceLoader;
+        this.meterRegistry = meterRegistry;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ArchivedJobMetadata getArchivedJobMetadata(
+        final String jobId
+    ) throws JobNotFoundException, JobNotArchivedException, JobDirectoryManifestNotFoundException {
+        final Instant startTime = Instant.now();
+        log.debug("Attempting to fetch archived job metadata for job {}", jobId);
+        final Set<Tag> tags = Sets.newHashSet();
+
+        try {
+            final String archiveLocation;
+
+            try {
+                archiveLocation = this.jobPersistenceService
+                    .getJobArchiveLocation(jobId)
+                    .orElseThrow(() -> new JobNotArchivedException("Job " + jobId + " wasn't archived"));
+            } catch (final GenieNotFoundException nfe) {
+                throw new JobNotFoundException(nfe);
+            }
+
+            final URI jobDirectoryRoot;
+
+            try {
+                jobDirectoryRoot = new URI(archiveLocation + SLASH).normalize();
+            } catch (final URISyntaxException e) {
+                throw new GenieRuntimeException("Unable to create URI from archive location: " + archiveLocation, e);
+            }
+
+            final URI manifestLocation;
+            if (StringUtils.isBlank(JobArchiveService.MANIFEST_DIRECTORY)) {
+                manifestLocation = jobDirectoryRoot.resolve(JobArchiveService.MANIFEST_NAME).normalize();
+            } else {
+                manifestLocation = jobDirectoryRoot
+                    .resolve(JobArchiveService.MANIFEST_DIRECTORY + SLASH)
+                    .resolve(JobArchiveService.MANIFEST_NAME)
+                    .normalize();
+            }
```

Comment:
Since this is coupled with JobArchiveService anyway (statically), why not just delegate this logic to it (eg. through a static helper method)?

Or maybe even better extract into a "manifest convention" thing that is used by both

Reply:
Yeah I'm not thrilled with this. I believe I left it in here purposely to refine it somehow in the future. right now with the code being shared still if I put it in a utility method it won't break anything when we move the `JobArchiveService` only into the Agent code base and we'll forget about this weak coupling. We need to do something better. Possibly some new field in the protocol that tells the agent where to store the archive metadata.